### PR TITLE
BUG: pivot_table sometimes returns Series (#4386)

### LINKED ISSUE #4386: DOC/BUG: pivot_table returns Series in specific circumstance
The docstrings and other documentation say that the `pivot_table` function **returns a DataFrame**.   However, this likely leads to confusion like #4371, because under narrow circumstances, passing a certain set of argument dtypes results in the function **returning a Series** (see ipython examples at end):
1. values is single string (not a list, not even a single valued list)
2. cols=None
3. aggfunc is single string/function (not a list, not even a single valued list)

Unfortunately, this is not clear from the docs or from normal use (except for condition 1).

Should this:
1. eventually be fixed to only return a DataFrame no matter the circumstances to be less confusing
2. be documented correctly (seems a little difficult to convey in the docstring and other docs without a lot of bulk).

My thoughts are changing the function to return only a DataFrame in future versions (> 0.13) and providing some deprecation warning in the meantime is better than trying to explain this in the docs.

I would be happy to provide the deprecation warning and document notes as a pull request.

Thanks.

``` python
Python 2.7.2 (v2.7.2:8527427914a2, Jun 11 2011, 15:22:34) 
Type "copyright", "credits" or "license" for more information.

IPython 1.0.dev -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.
%guiref   -> A brief reference about the graphical user interface.

In [1]: import pandas as pd
   ...: import numpy as np
   ...: pd.__version__
   ...: 
Out[1]: '0.12.0-57-g7bf2a7d'

In [2]: df = pd.DataFrame({'col1': [3, 4, 5], 'col2': ['C', 'D', 'E'], 'col3': [1, 3, 9]})

In [3]: df
Out[3]: 
   col1 col2  col3
0     3    C     1
1     4    D     3
2     5    E     9

In [4]: # Case 1: (a) values is single string label (b) cols is unspecified
   ...: #         (c) aggfunc is single lable/function (not a list)
   ...: # Expect: Series type
   ...: pivoted_1 = df.pivot_table('col1', rows=['col3', 'col2'], aggfunc=np.sum)
   ...: print pivoted_1
   ...: print type(pivoted_1)
col3  col2
1     C       3
3     D       4
9     E       5
Name: col1, dtype: int64
<class 'pandas.core.series.Series'>

In [5]: # Case 2: (a) values is single string label (b) cols is single string label
   ...: # Expected: DataFrame
   ...: pivoted_2 = df.pivot_table('col1', rows='col3', cols='col2', aggfunc=np.sum)
   ...: print pivoted_2
   ...: print type(pivoted_2)
col2   C   D   E
col3            
1      3 NaN NaN
3    NaN   4 NaN
9    NaN NaN   5
<class 'pandas.core.frame.DataFrame'>

In [6]: # Case 3: (a) values is single string label (b) cols is unspecified
   ...: #         (c) aggfunc is a list
   ...: # Expect: DataFrame
   ...: pivoted_3 = df.pivot_table('col1', rows='col3', aggfunc=[np.sum])
   ...: print pivoted_3
   ...: print type(pivoted_3)
      sum
col3     
1       3
3       4
9       5
<class 'pandas.core.frame.DataFrame'>
```

### PR DESCRIPTION
**Pull request by mroeschke**
- [x] closes #4386
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

The [`pivot_table` docs](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.pivot_table.html) mentions that a `DataFrame` should be returned. In some cases, a `Series` is returned. I had to also alter some existing tests that expected `pivot_table` to return a `Series`

Any feedback welcomed, thanks!

### CODE REVIEW

File: pandas/tools/pivot.py [hunk 1]
```diff
@@ -170,6 +170,9 @@ def pivot_table(data, values=None, index=None, columns=None, aggfunc='mean',
     if len(index) == 0 and len(columns) > 0:
         table = table.T
 
+    if isinstance(table, Series):
```
Comment (jreback) [inline pandas/tools/pivot.py]: can you see if you can find the actual root cause here?

e.g. may be as simple as making sure that the column/index seletors are list-like and not scalars

File: pandas/tools/tests/test_pivot.py [hunk 2]
```diff
@@ -350,7 +350,7 @@ def _check_output(result, values_col, index=['A', 'B'],
         # no rows
         rtable = self.data.pivot_table(columns=['AA', 'BB'], margins=True,
                                        aggfunc=np.mean)
-        tm.assertIsInstance(rtable, Series)
+        tm.assertIsInstance(rtable, DataFrame)
```
Comment (jreback) [inline pandas/tools/tests/test_pivot.py]: can you compare this with an expected result.

File: pandas/tools/tests/test_pivot.py [hunk 3]
```diff
@@ -485,7 +485,7 @@ def test_margins_no_values_no_cols(self):
         # Regression test on pivot table: no values or cols passed.
         result = self.data[['A', 'B']].pivot_table(
             index=['A', 'B'], aggfunc=len, margins=True)
-        result_list = result.tolist()
+        result_list = result.values.flatten().tolist()
         self.assertEqual(sum(result_list[:-1]), result_list[-1])
```
Comment (jreback) [inline pandas/tools/tests/test_pivot.py]: same here (I know this is old / not yours) .....

File: pandas/tools/pivot.py (hunk 1 above)
Comment (mroeschke) [inline pandas/tools/pivot.py]: Thanks @jreback. I found the root cause ([here](https://github.com/pandas-dev/pandas/blob/master/pandas/tools/pivot.py#L168)). I'll work on this and the tests tonight.

Once fixed, do you still want to keep the `isinstance(table, Series)` check or should I drop it?

**Comment by mroeschke** [comment]
Hi @jreback:

I added a fix to address the original bug. However, I believe `to_frame()` will be necessary because two scenarios ([pivoting with only the columns argument](https://github.com/pandas-dev/pandas/blob/master/pandas/tools/tests/test_pivot.py#L351) and [pivoting with no values or cols passed](https://github.com/pandas-dev/pandas/blob/master/pandas/tools/tests/test_pivot.py#L486)) will always return `Series` from the `groupby` and `unstack` operations respectively.

File: pandas/tools/pivot.py (hunk 1 above)
Comment (jreback) [inline pandas/tools/pivot.py]: actually I think a you can simply always make ``aggfunc`` a list (if its not None) and this will just work.

Comment (mroeschke) [inline pandas/tools/pivot.py]: Your solution works for the specific scenario in #4386, but a `Series`is still returned regardless when [pivoting with only the columns argument](https://github.com/pandas-dev/pandas/blob/master/pandas/tools/tests/test_pivot.py#L351) or [pivoting with no values or cols passed](https://github.com/pandas-dev/pandas/blob/master/pandas/tools/tests/test_pivot.py#L486) due to the `groupby`  and `unstack` operations respectively. In these cases, I think `to_frame()` might be necessary if `pivot_table` should always return a `DataFrame`.

Or should I just fix the specific scenario in #4386 and change the docs to mention that `pivot_table` can return a `Series`?

**Comment by jreback** [comment]
this looks like its duplicating #13554. Can you see if you tests are any different (maybe there is another case).

**Comment by mroeschke** [comment]
Ah yes, I wanted to move #13554 along since it looked inactive for a while (but it looks active now). I can look into adding more tests and building upon #13554.

**Comment by jreback** [comment]
can you rebase

**Comment by mroeschke** [comment]
Closing since this PR duplicated #13554 when it was inactive. Will reopen if #13554 stalls.

**Pull request closed without merging**